test/tests/functional/pbs_node_rampdown_keep_select.py

```diff
@@ -0,0 +1,1398 @@
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+from os import name
+from collections import Counter
+from copy import deepcopy
+
+
+class n_conf:
+    """
+    used to define node configuration info
+    """
+    def __init__(self, node_a={}, node_ct=0, usenatvn=False):
+        self.vnode_ct = node_ct
+        self.a = node_a
+        self.usenatvnode = usenatvn
+
+
+class new_res:
+    """
+    used to define new custom resource
+    """
+    def __init__(self, res_name, res_a={}):
+        self.res_name = res_name
+        self.a = res_a
+
+
+class test_config:
+    """
+    used to store config of a test case
+    """
+    def __init__(self, qsub_sel, keep_sel, sched_sel, expected_res, job_stat,
+                 rel_user, qsub_sel_after, sched_sel_after, job_stat_after,
+                 expected_res_after, skip_vnode_status_check=False,
+                 use_script=False):
+        self.qsub_sel = qsub_sel
+        self.keep_sel = keep_sel
+        self.sched_sel = sched_sel
+        self.expected_res = expected_res
+        self.job_stat = job_stat
+        self.rel_user = rel_user
+        self.qsub_sel_after = qsub_sel_after
+        self.sched_sel_after = sched_sel_after
+        self.job_stat_after = job_stat_after
+        self.expected_res_after = expected_res_after
+        self.skip_vnode_status_check = skip_vnode_status_check
+        self.use_script = use_script
+
+
+@requirements(num_moms=5)
+class TestPbsNodeRampDownKeepSelect(TestFunctional):
+    """
+    This tests the Node Rampdown Feature's extension called keep_select ,
```

Comment:
Space before comma.

Reply:
Removed in commit https://github.com/PBSPro/pbspro/pull/1452/commits/8f309272ad00f7587c830cad07238cb40598e24a